src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -545,30 +545,42 @@ protected virtual async Task<SecurityToken> ValidateSignedHttpRequestPayloadAsyn
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateM)
+            if (ShouldValidate(jwtSignedHttpRequest, signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateM, SignedHttpRequestClaimTypes.M))
                 ValidateMClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateU)
+            if (ShouldValidate(jwtSignedHttpRequest, signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateU, SignedHttpRequestClaimTypes.U))
                 ValidateUClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateP)
+            if (ShouldValidate(jwtSignedHttpRequest, signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateP, SignedHttpRequestClaimTypes.P))
                 ValidatePClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateQ)
+            if (ShouldValidate(jwtSignedHttpRequest, signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateQ, SignedHttpRequestClaimTypes.Q))
                 ValidateQClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateH)
+            if (ShouldValidate(jwtSignedHttpRequest, signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateH, SignedHttpRequestClaimTypes.H))
                 ValidateHClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateB)
+            if (ShouldValidate(jwtSignedHttpRequest, signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateB, SignedHttpRequestClaimTypes.B))
                 ValidateBClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
             return jwtSignedHttpRequest;
         }
 
+        /// <summary>
+        /// Determine if validation of a claim should happen.
+        /// </summary>
+        /// <param name="jwtSignedHttpRequest">The request being considered to validate the claim on.</param>
+        /// <param name="validateClaim">Force validation to always occur.</param>
+        /// <param name="claimName">The name of the claim to validate.</param>
+        /// <returns>Whether the given claim should be validated</returns>
+        internal virtual bool ShouldValidate(JsonWebToken jwtSignedHttpRequest, bool validateClaim, string claimName)
+        {
+            return validateClaim || (jwtSignedHttpRequest.TryGetClaim(claimName, out var claimValue) && claimValue != null);
```

Comment:
This is a change that may affect users where different clients are running different versions.
Suppose an older client is sending a value that is causing issues, the receiver currently has it turned off (validateClaim false). That client will succeed. With this change, that client will now start to fail.